File: amazon/139_word_break.cpp


// Given a string s and a dictionary of strings wordDict, 
// return true if s can be segmented into a space-separated 
// sequence of one or more dictionary words.

// Note that the same word in the dictionary may be reused multiple 
// times in the segmentation.

 

// Example 1:
// Input: s = "leetcode", wordDict = ["leet","code"]
// Output: true
// Explanation: Return true because "leetcode" can be segmented as "leet code".
// Example 2:

// Input: s = "applepenapple", wordDict = ["apple","pen"]
// Output: true
// Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
// Note that you are allowed to reuse a dictionary word.
// Example 3:

// Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
// Output: false

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.size()+1, false);
        dp[0] = true;
        
        for (int i = 1; i <= s.size();i++)
            for (int j = 0; j < i ; j++)
                if((dp[j]) && (find(wordDict.begin(), wordDict.end(), s.substr(j, i-j)) != wordDict.end())){
                    dp[i] = true;
                    break;
                    
                }
            return dp.back();

    }
};

//memoized dp 
class Solution{
private:
    class Tmort{
    public:
        char data;
        Tmort *child[26];
        bool point_end;

        Tmort(char ch){
            this->data = ch;
            for(int i =0 ; i < 26; i++){
                child[i] = NULL;
            }
            this->point_end = false;
        }
    };
    void addNodeT(string word, Tmort* root, int idx){
        if(idx >= word.length()){
            root->point_end = true;
            return;
        }
        char ch = word[idx]; 
        int index = ch - 'a';
        if(root->child[index] == NULL){
            root->child[index] = new Tmort(ch);
        }
        root = root->child[index];
        addNodeT(word, root, idx + 1); 
    }
    Tmort *org = NULL;
    unordered_map<int, bool> memo;
    bool solve(Tmort* root, string s , int start){
        if(start >= s.length()){
            return true;
        }
        if(memo.find(start) != memo.end()){
            return memo[start];
        }
        for(int i = start; i < s.length(); i++){
            char ch = s[i];
            int idx = ch - 'a';

            if(root->child[idx]==NULL){
                memo[start]= false;
                return false;
            }
            root = root->child[idx];
            if(root->point_end){
                bool x = solve(org, s ,i+1);
                if(x){
                    memo[start] = true;
                    return true;
                }
            }
        }
        memo[start] = false;
        return false;
    }
public:
    bool wordBreak(string s, vector<string>& wordDict){
        Tmort *root = new Tmort('*');
        org = root;
        bool ans = false;
        for (auto it : wordDict){
            addNodeT(it, root, 0);
        }
        ans = solve(root, s, 0);
        return ans;
    }
};

File: amazon/139_word_break.java


class Solution{
    public boolean wordBreak(String s, List<String> wordDict){
        boolean[] f = new boolean[s.length()+1];
        f[0] = true;

        for(int i = 1; i <= s.length();i++){
            for(int j = 0; j < i; j++){
                if(f[j] && wordDict.contains(s.substring(j, i))){
                    f[i] = true;
                    break;
                }
            }
        }
        return f[s.length()];
    }
}


//memoized
class Trim{
    Trim[] ch = new Trim[26];
    boolean point_end;
}
class Solution{
    public boolean wordBreak(String s, List<String> wordDict){
        int[] dp = new int[s.length()+1];
        Arrays.fill(dp, -1);
        Trim root = new Trim();
        for(String str:wordDict) insert(root, str);

        return dfs(0, s,root, s.length(),dp);
    }
    public boolean dfs(int idx, String s, Trim root, int n, int[] dp){
        if(idx >= n) return true;
        if(dp[idx] != -1) return dp[idx] == 0 ? false:true;
        Trim curr = root;
        for(int i = idx; i < n; i++){
            int j = s.charAt(i)-97;
            if(curr.ch[j]==null){
                dp[idx]=0; return false;
            }
            else{
                curr = curr.ch[j];
                if(curr.point_end){
                    if(dfs(i+1,s,root,n,dp)) {
                        dp[idx]=1; return true;
                    }
                }
            }
        }
        dp[idx] = 0;
        return false;
    }
    public void insert(Trim root, String s){
        Trim curr = root;
        for(int i = 0; i < s.length();i++){
            int idx = s.charAt(i)-97;
            if(curr.ch[idx]==null){
                curr.ch[idx]= new Trim();
            }
            curr = curr.ch[idx];
        }
        curr.point_end = true;
    }
}